articles/best-practices/rules.md

```diff
@@ -1,247 +1,565 @@
+
+Removing calls to the Management API (as well as the extra call required to get the appropriate <dfn data-key="access-token">Access Token</dfn>) will make your rule code perform better as well as be more reliable. 
+
+### Consider use of explicit timeouts when making API calls
+
+When calling APIs or accessing external services consider specifiying explicit timeout(s). The specific timeout value you choose will typically vary based on your user case, but in general, choosing a one that's as low as possible - whilst bearing in mind the performance charisteristics of the external service - is advised.
+
+::: panel Best Practice
+Whether you choose to employ use of explicit timeouts, or you choose to opt for implicit timout processing, always be sure to cater for [error](#error-handling) and/or [exception](#exceptions) conditions that may occur as a rusult of any timeout period expiration.
+:::
+
+## Security
+
+### Use HTTPS
+
+Always use HTTPS, not HTTP, when making calls to external services or when executing [redirect](#redirection) as part of your rule implementation.
+
+### Store security sensitive values in rule Settings
+
+Security sensitive information, such as credentials or API keys, should be stored in your [rule settings](${manage_url}/#/rules) where they'll be obfuscated, encrypted and available via the [`configuration`](#configuration-object) object. Do not store these values as literals in your rules code. For example, do not write code like this:
+
+```js
+const myApiKey = 'abc123';
+```
+
+Instead, prefer to store (secret) information so that it's accesible via the [`configuration`](#configuration-object) object:
 
 ```js
-context.multi-factor = {
-  provider: 'guardian',
-  allowRememberBrowser: true
-};
+const myApiKey = configuration.myApiKey;
 ```
 
-### Always check if an email is verified
+### Don’t send entire [`context`](#context-object) object to external services
+
+For rules that send information to an external service, make sure you are not sending the entire [context](#context-object), since this object may contain tokens or other sensitive data. For rules that send information to external services, you should only send a *subset* of the less sensitive attributes from the `context` object when and where necessary.
 
-Whenever granting authorization based on an email address, always start by checking if the user has verified the email address:
+::: warning
+In a similar fashion, avoid passing **any** aspect of the [`auth0`](#auth0-object) object outside of a rule too.  
```

Comment:
```suggestion
In a similar fashion, avoid passing **any** aspect of the [`auth0`](#auth0-object) object outside of a rule.  
```

Reply:
I'm going to leave the "too" as this sentence is about an object "in addition" to `context`